### ISSUE TITLE
seL4 failed assertion 'sc->scRefillHead < sc->scRefillMax'

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#633]
Not sure what's going on, but since today occasionally I'm getting:
```
seL4 failed assertion 'sc->scRefillHead < sc->scRefillMax' at /home/iz/vmm-mep/kernel/src/kernel/sporadic.c:128 in function refill_pop_head
halting...
Kernel entry via Interrupt, irq 0
```
It may be triggered by application code calling seL4_DebugPutString() though.

seL4 commit 561c6e65bef2d9, SMP, MCS and HYP configured.

I never got this with an older kernel, it seems something fell over since June / commit db7bb413.

**Comment by Indanz** [comment]
The only other place that sets ksCurThread is switchToThread(), which is only called by schedule(), either directly or indirectly, and schedule() always calls switchSchedContext().

switchToIdleThread() is only called via schedule() or by ipiStallCoreCallback(), so all in all ipiStallCoreCallback() is the only call path which changes the current thread without updating the ksCurSC.

Is the above fix sufficient, or should ipiStallCoreCallback() do more, like call switchSchedContext()?

**Comment by lsf37** [comment]
Thanks for debugging this, `sc->scRefillHead < sc->scRefillMax` is definitely supposed to not fail. @pingerino or @kent-mcleod might be able to say if we need `switchSchedContext()` here or if the assignment is enough.

@mktnk3 and @michaelmcinerney anything from the verification side on this one? We don't see the `ipiStallCoreCallback()` code path in the verification, because that is SMP only, but I think we have an invariant about `ksCurSC` and `ksCurThread` that might be relevant here.

**Comment by pingerino** [comment]
IIRC you don't need swtichSchedContext when switching to the idle thread (the idle thread always has budget and isn't doing anything fancy with thread states, nor can it donate its scheduling context). 

Are we still running the SMP + MCS version of sel4test? what about sel4bench? one of these should have caught this, and if not, we should add a test case. 

Thanks for reporting and investigating!

**Comment by michaelmcinerney** [comment]
> @mktnk3 and @michaelmcinerney anything from the verification side on this one? We don't see the `ipiStallCoreCallback()` code path in the verification, because that is SMP only, but I think we have an invariant about `ksCurSC` and `ksCurThread` that might be relevant here.

We do have `0 < scRefillMax sc --> scRefillHead sc < scRefillMax sc` as part of `valid_sched_context'`, and we have already proved that `refillBudgetCheck` preserves `invs'` (it does include an assert that the sc is active, but this seems fine). So it seems more likely that this issue is coming from the SMP code.

`valid_idle` says that the idle thread and idle sc are linked. Is that the invariant you were thinking of? One other is `cur_sc_tcb`, which is part of `invs`, but I'm not sure that's relevant here.

**Comment by lsf37** [comment]
> Are we still running the SMP + MCS version of sel4test? what about sel4bench? one of these should have caught this, and if not, we should add a test case.

Yes, both are running, but neither caught it. A test case for it would definitely be very good.

**Comment by Indanz** [comment]
> IIRC you don't need swtichSchedContext when switching to the idle thread (the idle thread always has budget and isn't doing anything fancy with thread states, nor can it donate its scheduling context).

True, but shouldn't the consumed time be charged to the previous SC instead of the idle thread? This doesn't matter if the SC goes away, but that usually doesn't happen. So at least the following portion of commitTime:
```
NODE_STATE(ksCurSC)->scConsumed += NODE_STATE(ksConsumed);
NODE_STATE(ksConsumed) = 0llu;
```
> Are we still running the SMP + MCS version of sel4test? what about sel4bench? one of these should have caught this, and if not, we should add a test case.

This only triggers with SMP + MCS and a periodic SC, so perhaps this corner case was forgotten. I will add a test for this and similar corner cases.

**Comment by Indanz** [comment]
> `valid_idle` says that the idle thread and idle sc are linked.

This is the one that gets broken currently.

**Comment by Indanz** [comment]
I've added a test for this to sel4test, but that was surprisingly hard to trigger the use-after-free: It seems to be an actual race between the SC delete and an end of budget timer IRQ. Usually the bug is masked by any schedule calls, and a schedule is triggered after deleting the active SC or TCB. All in all I do not know how to add a cross-platform test that's guaranteed to make sense.

Currently I have the below, which works with budget up to 30 us, but only with a debug kernel. If RELEASE is true it doesn't crash the kernel:
```
/* Try to recreate race situation of issue 633 */
int test_smp_delete_sched_context(env_t env)
{
    seL4_Error error;
    helper_thread_t helper;
    volatile int state = 0;
    int prev_state = state;

    create_helper_thread(env, &helper);
    set_helper_priority(env, &helper, env->priority - 1);
    /* Run it on another core and time it such that it runs out of budget during or just after the SC free */
    error = seL4_SchedControl_Configure(simple_get_sched_ctrl(&env->simple, CONFIG_MAX_NUM_NODES - 1),
            helper.thread.sched_context.cptr, 20, 10 * US_IN_S, 0, 0);
    test_eq(error, seL4_NoError);

    start_helper(env, &helper, (helper_fn_t)sched_context_0005_helper_fn, (seL4_Word)&state, 0, 0, 0);

    /* Wait till helper runs */
    while (state == prev_state);

    vka_free_object(&env->vka, &helper.thread.sched_context);

    return sel4test_get_result();
}
DEFINE_TEST(SCHED_CONTEXT_0014, "Test deleting a scheduling context while bound tcb runs on another core",
            test_smp_delete_sched_context, config_set(CONFIG_KERNEL_MCS) &&
            config_set(CONFIG_HAVE_TIMER) && CONFIG_MAX_NUM_NODES > 1);
```

**Comment by lsf37** [comment]
That is excellent, it only needs to trigger the bug in one of the configurations.

I checked `sel4bench` again, and contrary to what I said above it does actually *not* run SMP + MCS (never did, it has a comment that this should still be implemented, and if I enable the combination the test driver crashes on ia32). It runs MCS and SMP only separately. Since SMP+MCS for sel4bench seems to work on the other platforms, this is probably something we should fix/add, but that is a separate story (I also haven't checked if the results make any sense for that combination).

**Comment by pingerino** [comment]
> True, but shouldn't the consumed time be charged to the previous SC instead of the idle thread? This doesn't matter if the SC goes away, but that usually doesn't happen. So at least the following portion of commitTime:

Correct - we don't bother charging it if the SC is being deleted. If it's possible that the SC is used again, it definitely needs to be charged. This is a good invariant to prove, (in addition to not charging an SC twice), and since we have no proofs of SMP code it could definitely be a bug.

**Comment by pingerino** [comment]
Also, awesome! Thanks for adding the test!

**Comment by Indanz** [comment]
Currently timing accounting is correct as it is updated when scheduling, which always happens because a schedule is triggered when the core is unstalled. Setting ksCurSC in switchToIdleThread() is the wrong thing to do, because that would mess this up. I don't think we can have a simple fix for this race without breaking time accounting, so we need a proper fix.

**Comment by lsf37** [comment]
> @lsf37 you meant sel4test I assume? SMP + MCS is tested by sel4test.

I meant `sel4bench` in this case because Anna was asking about both earlier. SMP+MCS does run in `sel4test`, but not `sel4bench`. We should probably also add it to the latter if we are at the stage where we would call this combination properly supported.

**Comment by Indanz** [comment]
Forgot to close this, it's fixed by merge of #636.
